fawltydeps/packages.py

```diff
@@ -317,14 +317,27 @@ class TemporaryPipInstallResolver(BasePackageResolver):
     def installed_requirements(
         venv_dir: Path, requirements: List[str]
     ) -> Iterator[Path]:
-        """Install the given requirements into venv_dir with `pip install`."""
+        """Install the given requirements into venv_dir with `pip install`.
+
+        If `pip install` fails, we will log a warning, then retry by attempting
+        to `pip install` each given requirement in order (logging any errors
+        that occur), before we finally enter the caller's context.
```

Comment:
```suggestion
```
Nit: Do you think this comment is needed? It duplicates exactly what the code does, which may cause some later code vs comments discrepancy if we change something.

Reply:
I would agree if this was a _comment_ in the code, but when it's a _docstring_ I think it is somewhat different:

The docstring is often visible in places where the code is not (or less) visible (see below), and I believe the behavior described in these three lines is valuable, even for a casual user of this method (that would not want to read the code), as it describes postconditions that the caller must be aware of (i.e. do we enter the context on failure? Do we abort of the first failing package, or do we persist/retry?).

Even if the code and docstring later would disagree (which of course should be fixed as soon as it is discovered), I think it is valuable to capture the _intended behavior_ of this method in as simple English as possible. Of course, the above text is not particularly good in that regard, and I think it could be simplified further:

```suggestion
        We try to install as many of the given requirements as possible. Failed
        requirements will be logged with warning messages, but no matter how
        many failures we get, we will still enter the caller's context.
```

---

To my point of docstrings being _more visible_ than code: For example, in the Python REPL you can use `help()` to view the signature + docstring of something to learn how to call it:

```sh
$ python
Python 3.10.9 (main, Dec  6 2022, 18:44:57) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from fawltydeps.packages import TemporaryPipInstallResolver
>>> help(TemporaryPipInstallResolver.installed_requirements)
Help on function installed_requirements in module fawltydeps.packages:

installed_requirements(venv_dir: pathlib.Path, requirements: List[str]) -> Iterator[pathlib.Path]
    Install the given requirements into venv_dir with `pip install`.
    
    If `pip install` fails, we will log a warning, then retry by attempting
    to `pip install` each given requirement in order (logging any errors
    that occur), before we finally enter the caller's context.
    
    It is up to the caller to handle any requirements that we failed
    to install.
```

Another example is when a Python library is documented via an automatically generated website; the docstrings are included, but the actual code is not included (or, at best, referenced in a link).

Reply:
Cleaned up the docstring in 41d9a42, it now reads:
```python
        """Install the given requirements into venv_dir with `pip install`.

        We try to install as many of the given requirements as possible. Failed
        requirements will be logged with warning messages, but no matter how
        many failures we get, we will still enter the caller's context. It is
        up to the caller to handle any requirements that we failed to install.
        """
```